docs/structured-streaming-programming-guide.md

```diff
@@ -2251,13 +2251,13 @@ When the streaming query is started, Spark calls the function or the object’s
 
 - The close() method (if it exists) is called if an open() method exists and returns successfully (irrespective of the return value), except if the JVM or Python process crashes in the middle.
 
-- **Note:** The partitionId and epochId in the open() method can be used to deduplicate generated data 
-  when failures cause reprocessing of some input data. This depends on the execution mode of the query. 
-  If the streaming query is being executed in the micro-batch mode, then every partition represented 
-  by a unique tuple (partition_id, epoch_id) is guaranteed to have the same data. 
-  Hence, (partition_id, epoch_id) can be used to deduplicate and/or transactionally commit 
-  data and achieve exactly-once guarantees. However, if the streaming query is being executed 
-  in the continuous mode, then this guarantee does not hold and therefore should not be used for deduplication.
+- **Note:** Spark does not guarantee same output for (partitionId, epochId) on failure, so deduplication
+  cannot be achieved with (partitionId, epochId). e.g. source provides different number of
+  partitions for some reasons, Spark optimization changes number of partitions, etc.
+  See [SPARK-28650](https://issues.apache.org/jira/browse/SPARK-28650) for more details. `epochId` can still be used
+  for deduplication, but there's less benefit to leverage this, as the chance for Spark to successfully write all
+  partitions and fail to checkpoint the batch is small. You also need to care about whether epoch is fully written,
+  via ensuring all partitions for the epochId are written successfully.
```

Comment:
> ensuring all partitions for the epochId are written successfully.

Hm, looks like `ForeachWriter` doesn't know the number of partitions, so it cannot implement something like this.

Reply:
Oh you're right. There's no context around this unless end users do some kind of hacks (aggregating over couple of batches) and it's not worth to do with such hard hacks. I'll just get rid of guarantee for deduplication on `epochId`.